File: 2010_2011_mf_sujet_4_convergences_de_processus/Ehrenfest3.m
function Y = Ehrenfest3(d,X)
%UNTITLED6 Simule un pas d'Ehrenfest
%   Detailed explanation goes here
U=rand(1);
if (U<d/(d+1))
    V=rand(1);
    Y=X +(V>(X/d))-(V<(X/d));
else Y=X;
end
end

File: 2010_2011_mf_sujet_4_convergences_de_processus/Ehrenfest3.py
import numpy as np


def Ehrenfest3(d, X):
    """Simulate one step of the lazy Ehrenfest chain (aperiodic version)."""
    U = np.random.rand()
    if U < d / (d + 1):
        V = np.random.rand()
        Y = X + (V > X / d) - (V < X / d)
    else:
        Y = X
    return Y
